File: scripts/eval/sample-oa-zip.py
#!/usr/bin/env python3
"""Reservoir-sample an OpenAddresses national CSV (inside a ZIP64 zip) into a resolver-eval JSONL.

The ingest-openaddresses.mjs path uses `unzip -p`, which chokes on ZIP64 (national datasets like FR
BAN have a >4GB CSV). Python's zipfile streams ZIP64 fine. Same output shape as the other
openaddresses-*-sample.jsonl files: {input, lat, lon, expected:{locality,region,postcode}, state, source}.

Usage:
  python3 scripts/eval/sample-oa-zip.py --zip /tmp/oa-cache/fr__countrywide.zip --country FR \
    --target 3000 --seed 42 --out data/eval/external/openaddresses-fr-sample.jsonl
"""
import argparse, csv, io, json, random, zipfile

BBOX = {
    "FR": (41.0, 51.5, -5.5, 9.7),
    "NL": (50.7, 53.6, 3.3, 7.3),
    "IT": (35.0, 47.2, 6.5, 18.6),
    "ES": (27.5, 43.9, -18.3, 4.4),  # mainland + Balearics + Canaries
}
# Normalize a raw CSV row to the OA-standard column names. Most national OA exports already use
# NUMBER/STREET/CITY/POSTCODE/REGION/LAT/LON; Spain ships the CartoCiudad schema (X/Y, nombre_via,
# poblacion, provincia, …) so we remap it. STREET folds in the Spanish street-type (CALLE/AVENIDA).
def normalize_row(country, r):
    if country == "ES":
        via = f"{(r.get('tipo_vial') or '').strip()} {(r.get('nombre_via') or '').strip()}".strip()
        return {"LAT": r.get("Y"), "LON": r.get("X"), "NUMBER": (r.get("numero") or "").strip(),
                "STREET": via, "CITY": (r.get("poblacion") or "").strip(),
                "POSTCODE": (r.get("cod_postal") or "").strip(), "REGION": (r.get("provincia") or "").strip()}
    return r

# Render order per locale (the raw string the parser sees).
def render(country, r):
    num, street, pc, city = r["NUMBER"].strip(), r["STREET"].strip(), r["POSTCODE"].strip(), r["CITY"].strip()
    if country == "FR":  # "12 Rue de Rivoli, 75001 Paris"
        return f"{num} {street}, {pc} {city}".strip(", ")
    if country in ("NL", "IT", "ES"):  # number after street: "Via Roma 12, 20121 Milano" / "Calle de Alcalá 1, 28014 Madrid"
        return f"{street} {num}, {pc} {city}".strip(", ")
    return f"{num} {street}, {pc} {city}"

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--zip", required=True)
    ap.add_argument("--country", required=True)
    ap.add_argument("--target", type=int, default=3000)
    ap.add_argument("--seed", type=int, default=42)
    ap.add_argument("--out", required=True)
    args = ap.parse_args()
    rng = random.Random(args.seed)
    minlat, maxlat, minlon, maxlon = BBOX[args.country]

    z = zipfile.ZipFile(args.zip)
    csv_name = [n for n in z.namelist() if n.endswith(".csv")][0]
    reservoir, seen = [], 0
    with z.open(csv_name) as raw:
        reader = csv.DictReader(io.TextIOWrapper(raw, encoding="utf-8", errors="replace"))
        for raw_row in reader:
            r = normalize_row(args.country, raw_row)
            try:
                lat, lon = float(r["LAT"]), float(r["LON"])
            except (ValueError, KeyError, TypeError):
                continue
            if not (minlat <= lat <= maxlat and minlon <= lon <= maxlon):
                continue
            city, pc = (r.get("CITY") or "").strip(), (r.get("POSTCODE") or "").strip()
            if not city or not pc:
                continue  # admin-level eval needs city + postcode
            seen += 1
            row = {
                "input": render(args.country, r),
                "lat": lat, "lon": lon,
                "expected": {"locality": city, "region": (r.get("REGION") or "").strip() or None, "postcode": pc},
                "state": args.country,
                "source": f"openaddresses:{args.country.lower()}/countrywide",
            }
            if len(reservoir) < args.target:
                reservoir.append(row)
            else:
                j = rng.randint(0, seen - 1)
                if j < args.target:
                    reservoir[j] = row
    with open(args.out, "w", encoding="utf-8") as f:
        for row in reservoir:
            f.write(json.dumps(row, ensure_ascii=False) + "\n")
    print(f"wrote {len(reservoir)} rows (sampled from {seen:,} valid in-bbox) → {args.out}")

if __name__ == "__main__":
    main()

File: scripts/eval/sample-oa-zip.ts
/**
 * @copyright Sister Software
 * @license AGPL-3.0
 * @author Teffen Ellis, et al.
 *
 *   Reservoir-sample an OpenAddresses national CSV (inside a ZIP64 zip) into a resolver-eval JSONL.
 *
 *   The ingest-openaddresses.mjs path uses `unzip -p`, which historically choked on ZIP64 (national
 *   datasets like FR BAN have a >4GB CSV); Debian's UnZip 6.00 has large-file support, and we
 *   stream the member through `csv-parse` (see scripts/lib/zip-csv.ts). Same output shape as the
 *   other openaddresses-*-sample.jsonl files: {input, lat, lon,
 *   expected:{locality,region,postcode}, state, source}.
 *
 *   Usage: node --experimental-strip-types scripts/eval/sample-oa-zip.ts --zip
 *   /tmp/oa-cache/fr__countrywide.zip\
 *   --country FR --target 3000 --seed 42 --out data/eval/external/openaddresses-fr-sample.jsonl
 *
 *   Ported faithfully from scripts/eval/sample-oa-zip.py. NOTE: the seeded reservoir RNG is
 *   distribution-faithful but NOT CPython-bit-identical (see python-random.ts); the algorithm,
 *   filters, and schema are preserved.
 */

import { writeFileSync } from "node:fs"
import { parseArgs } from "node:util"

import { pyJsonDumps } from "../lib/python-json.ts"
import { SeededRandom } from "../lib/python-random.ts"
import { csvRecordsFromZip, firstCSVEntry } from "../lib/zip-csv.ts"

type Row = Record<string, string | undefined>

const BBOX: Record<string, [number, number, number, number]> = {
	FR: [41.0, 51.5, -5.5, 9.7],
	NL: [50.7, 53.6, 3.3, 7.3],
	IT: [35.0, 47.2, 6.5, 18.6],
	ES: [27.5, 43.9, -18.3, 4.4], // mainland + Balearics + Canaries
}

// Normalize a raw CSV row to the OA-standard column names. Most national OA exports already use
// NUMBER/STREET/CITY/POSTCODE/REGION/LAT/LON; Spain ships the CartoCiudad schema (X/Y, nombre_via,
// poblacion, provincia, …) so we remap it. STREET folds in the Spanish street-type (CALLE/AVENIDA).
function normalizeRow(country: string, r: Row): Row {
	if (country === "ES") {
		const via = `${(r.tipo_vial ?? "").trim()} ${(r.nombre_via ?? "").trim()}`.trim()

		return {
			LAT: r.Y,
			LON: r.X,
			NUMBER: (r.numero ?? "").trim(),
			STREET: via,
			CITY: (r.poblacion ?? "").trim(),
			POSTCODE: (r.cod_postal ?? "").trim(),
			REGION: (r.provincia ?? "").trim(),
		}
	}

	return r
}

/** Python `str.strip(", ")` — strip leading/trailing chars that are comma or space. */
function stripCommaSpace(s: string): string {
	return s.replace(/^[, ]+/, "").replace(/[, ]+$/, "")
}

// Render order per locale (the raw string the parser sees).
function render(country: string, r: Row): string {
	const num = (r.NUMBER ?? "").trim()
	const street = (r.STREET ?? "").trim()
	const pc = (r.POSTCODE ?? "").trim()
	const city = (r.CITY ?? "").trim()

	if (country === "FR") return stripCommaSpace(`${num} ${street}, ${pc} ${city}`)

	// "12 Rue de Rivoli, 75001 Paris"
	// number after street: "Via Roma 12, 20121 Milano" / "Calle de Alcalá 1, 28014 Madrid"
	if (country === "NL" || country === "IT" || country === "ES")
		return stripCommaSpace(`${street} ${num}, ${pc} ${city}`)

	return `${num} ${street}, ${pc} ${city}`
}

/**
 * Python `float(s)`: empty/non-numeric -> null (covers the ValueError/KeyError/TypeError -> continue).
 */
function pyFloat(s: string | undefined): number | null {
	if (s == null) return null
	const t = s.trim()

	if (t === "") return null
	const v = Number(t)

	return Number.isNaN(v) ? null : v
}

async function main(): Promise<void> {
	const { values } = parseArgs({
		options: {
			zip: { type: "string" },
			country: { type: "string" },
			target: { type: "string", default: "3000" },
			seed: { type: "string", default: "42" },
			out: { type: "string" },
		},
	})

	for (const req of ["zip", "country", "out"] as const) {
		if (!values[req]) {
			process.stderr.write(`error: the following arguments are required: --${req}\n`)
			process.exit(2)
		}
	}
	const zip = values.zip!
	const country = values.country!
	const target = Number(values.target)
	const out = values.out!

	const rng = new SeededRandom(Number(values.seed))
	const bbox = BBOX[country]

	if (!bbox) throw new Error(`no bbox for country ${country}`)
	const [minlat, maxlat, minlon, maxlon] = bbox

	const csvName = firstCSVEntry(zip)
	const reservoir: Record<string, unknown>[] = []
	let seen = 0

	for await (const rawRow of csvRecordsFromZip(zip, csvName)) {
		const r = normalizeRow(country, rawRow)
		const lat = pyFloat(r.LAT)
		const lon = pyFloat(r.LON)

		if (lat === null || lon === null) continue

		if (!(minlat! <= lat && lat <= maxlat! && minlon! <= lon && lon <= maxlon!)) continue
		const city = (r.CITY ?? "").trim()
		const pc = (r.POSTCODE ?? "").trim()

		if (!city || !pc) continue // admin-level eval needs city + postcode
		seen += 1
		const region = (r.REGION ?? "").trim() || null
		const row = {
			input: render(country, r),
			lat,
			lon,
			expected: { locality: city, region, postcode: pc },
			state: country,
			source: `openaddresses:${country.toLowerCase()}/countrywide`,
		}

		if (reservoir.length < target) {
			reservoir.push(row)
		} else {
			const j = rng.randint(0, seen - 1)

			if (j < target) reservoir[j] = row
		}
	}

	writeFileSync(out, reservoir.map((row) => pyJsonDumps(row, { ensureAscii: false }) + "\n").join(""))
	console.log(`wrote ${reservoir.length} rows (sampled from ${seen.toLocaleString("en-US")} valid in-bbox) → ${out}`)
}

await main()
